slatedb/src/compactions_store.rs

```diff
@@ -0,0 +1,495 @@
+use crate::transactional_object::{
+    DirtyObject, FenceableTransactionalObject, MonotonicId, SequencedStorageProtocol,
+    SimpleTransactionalObject, TransactionalObject, TransactionalStorageProtocol,
+};
+use chrono::Utc;
+use object_store::path::Path;
+use object_store::ObjectStore;
+use serde::Serialize;
+#[cfg(test)]
+use std::ops::RangeBounds;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Represents the compactions stored in the object store. This type tracks the current
+/// contents and id of the stored compactions state, and allows callers to read and update
+/// it using transactional semantics.
+pub(crate) struct StoredCompactions {
+    inner: SimpleTransactionalObject<Compactions>,
+}
+
+impl StoredCompactions {
+    async fn init(
+        store: Arc<CompactionsStore>,
+        compactions: Compactions,
+    ) -> Result<Self, SlateDBError> {
+        let inner = SimpleTransactionalObject::<Compactions>::init(
+            Arc::clone(&store.inner)
+                as Arc<dyn TransactionalStorageProtocol<Compactions, MonotonicId>>,
+            compactions.clone(),
+        )
+        .await?;
+        Ok(Self { inner })
+    }
+
+    /// Create a new compactions record with the supplied compactor epoch and no compactions.
+    pub(crate) async fn create(
+        store: Arc<CompactionsStore>,
+        compactor_epoch: u64,
+    ) -> Result<Self, SlateDBError> {
+        let compactions = Compactions::new(compactor_epoch);
+        Self::init(store, compactions).await
+    }
+
+    /// Load the current compactions state from the supplied compactions store. If there is no
+    /// compactions state stored, this fn returns None. Otherwise, on success it returns a
+    /// Result with an instance of StoredCompactions.
+    pub(crate) async fn try_load(
+        store: Arc<CompactionsStore>,
+    ) -> Result<Option<Self>, SlateDBError> {
+        let Some(inner) =
+            SimpleTransactionalObject::<Compactions>::try_load(Arc::clone(&store.inner)
+                as Arc<dyn TransactionalStorageProtocol<Compactions, MonotonicId>>)
+            .await?
+        else {
+            return Ok(None);
+        };
+        Ok(Some(Self { inner }))
+    }
+
+    /// Load the current compactions state from the supplied compactions store. If successful,
+    /// this method returns a [`Result`] with an instance of [`StoredCompactions`].
+    /// If no compactions could be found, the error [`LatestTransactionalObjectVersionMissing`] is returned.
+    #[cfg(test)]
+    pub(crate) async fn load(store: Arc<CompactionsStore>) -> Result<Self, SlateDBError> {
+        SimpleTransactionalObject::<Compactions>::try_load(Arc::clone(&store.inner)
+            as Arc<dyn TransactionalStorageProtocol<Compactions, MonotonicId>>)
+        .await?
+        .map(|inner| Self { inner })
+        .ok_or(LatestTransactionalObjectVersionMissing)
+    }
+
+    #[allow(dead_code)]
```

Comment:
The `dead_code` and `test` annotations in this file should go away once we implement the follow-on PR (use the GC to read the `Compactions` from object storage).